### PR TITLE
[JAX FE]: Support jax.lax.gather operation for JAX

### LINKED ISSUE #26575: [Good First Issue][JAX FE]: Support jax.lax.gather operation for JAX
### Context

OpenVINO component responsible for support of JAX/Flax models is called as JAX Frontend (JAX FE). JAX FE converts a JAX/Flax model represented by `ClosedJAXpr` graph object with operations from [jax.lax opset](https://jax.readthedocs.io/en/latest/jax.lax.html) to OpenVINO IR containing operations from [OpenVINO opset](https://docs.openvino.ai/2024/documentation/openvino-ir-format/operation-sets/available-opsets/opset14.html).

In order to infer JAX/Flax models containing [jax.lax.gather](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.gather.html) operation by OpenVINO, JAX FE needs to be extended with this operation support.

### What needs to be done?

For [ jax.lax.gather](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.gather.html) operation support, you need to implement the corresponding loader into JAX FE [op](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/jax/src/op) directory and to register it into the dictionary of [Loaders](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/jax/src/op_table.cpp). One loader is responsible for conversion (or decomposition) of one type of JAX operation.

Here is an example of loader implementation for [jax.lax.reshape](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.reshape.html) operation:

```cpp
OutputVector translate_reshape(const NodeContext& context) {
    num_inputs_check(context, 1, 1);
    Output<Node> input = context.get_input(0);
    auto new_sizes = context.const_named_param<std::vector<int64_t>>("new_sizes");
    if (context.has_param("dimensions")) {
        auto dimensions = context.const_named_param<std::vector<int64_t>>("dimensions");
        // transpose the input first.
        auto permutation_node = std::make_shared<v0::Constant>(element::i64, Shape{dimensions.size()}, dimensions);
        input = std::make_shared<v1::Transpose>(input, permutation_node);
    }

    auto new_shape_node = std::make_shared<v0::Constant>(element::i64, Shape{new_sizes.size()}, new_sizes);
    Output<Node> res = std::make_shared<v1::Reshape>(input, new_shape_node, false);
    return {res};
};
```

In this example, `translate_reshape` expresses `jax.lax.reshape` using OpenVINO opset. Since `jax.lax.reshape` performs transposition and tensor reshaping according to JAX documentation, the resulted decomposition contains OpenVINO `Transpose`  and `Reshape` operations. For `Transpose` and `Reshape` nodes, this conversion parses constant parameters `dimensions` to permute input tensor and `new_size` that is the target shape of the result.

Once you are done with implementation of the translator, you need to implement the corresponding layer tests `test_gather.py` and put it into [layer_tests/jax_tests](https://github.com/openvinotoolkit/openvino/tree/master/tests/layer_tests/jax_tests) directory. Example how to run some layer test:

```
export TEST_DEVICE=CPU
export JAX_TRACE_MODE=JAXPR
export 
cd openvino/tests/layer_tests/jax_tests
pytest test_reshape.py
```

### Example Pull Requests

- https://github.com/openvinotoolkit/openvino/pull/26288
- https://github.com/openvinotoolkit/openvino/pull/26254

### Resources

- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [How to Build OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/docs/dev/build.md)
- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Intel DevHub Discord channel](https://discord.gg/7pVRxUwdWG) - engage in discussions, ask questions and talk to OpenVINO developers

### Contact points

- @openvinotoolkit/openvino-jax-frontend-maintainers 
- @rkazants in GitHub and Discord

### Ticket

_No response_

### PR DESCRIPTION
**Pull request by Jyc323**
### Details:
 - implement translation in gather.cpp
 - register function in op_table.cpp
 - Add pytest test case in test_gather.py

### Tickets:
 - Closes #26575

### CODE REVIEW

**Comment by rkazants** [comment]
build_jenkins

**Comment by Jyc323** [comment]
Hi @rkazants , I may need some help. I modified the code, and while the previous error is now resolved, I got some new errors during model conversion:

```
self = <FrontEnd 'tf'>
model = <openvino._pyopenvino.InputModel object at 0x7f575d8cce90>

    def convert(self, model: Union[Model, InputModel]) -> Model:
>       converted_model = super().convert(model)
E       openvino._pyopenvino.OpConversionFailure: Check 'is_conversion_successful' failed at src/frontends/tensorflow/src/frontend.cpp:478:
E       FrontEnd API failed with OpConversionFailure:
E       [TensorFlow Frontend] Internal error, no translator found for operation(s): XlaGather
E       To facilitate the conversion of unsupported operations, refer to Frontend Extension documentation: https://docs.openvino.ai/latest/openvino_docs_Extensibility_UG_Frontend_Extensions.htmlCan you help me check if there is something 
```
I'm wondering if there’s anything I can do to fix this on my end, or if it’s an issue with the OpenVINO framework itself. I've also uploaded the full log for reference.

Any help would be greatly appreciated—thank you in advance!
[gather_v2.log](https://github.com/user-attachments/files/19410293/gather_v2.log)

**Comment by Jyc323** [comment]
@p-wysocki @rkazants I renamed the parameter. please review and re-build, thank you!

**Comment by mlukasze** [comment]
build_jenkins

**Comment by Jyc323** [comment]
@mlukasze I fixed the error "No param id corresponding name exists: axes" in layer test. Please review and let me know if there is any issue or suggestion. Thank you!

**Comment by Jyc323** [comment]
Hi @mlukasze,

I made a small update to `gather.cpp` in the JAX frontend. When running the JAX layer tests, 2026 tests passed, but 12 tests from `test_gather.py` failed.

The error was due to `context.const_named_param("axis")` failing, since the `axis` parameter is not always passed as a named constant. In some cases, it appears to be stored in `start_index_map` instead.

To handle both cases, I added a `try-catch` block in `gather.cpp`:
- Try to get `axis` from `const_named_param("axis")`
- If it fails, fall back to reading it from `start_index_map`

I'm not entirely sure if this is the correct or intended behavior, so I'd appreciate your verification.

**Comment by mlukasze** [comment]
build_jenkins
